### ISSUE TITLE
`CairoMakie`: polygon with hole example not working

### ISSUE THREAD

**Issue by fatteneder** [MakieOrg/Makie.jl#2456]
See the 2nd example at the docs page: https://docs.makie.org/v0.18.3/examples/plotting_functions/poly/index.html

Expected output: 
![image](https://user-images.githubusercontent.com/26469701/204379640-cac14442-d704-40df-9d8a-bc1050adf68f.png)

**Comment by jaakkor2** [comment]
This polygon works in v0.18.3, i.e., with reversed winding of the hole
``´
p = Polygon(
    Point2f[(0, 0), (2, 0), (3, 1), (1, 1)],
    [Point2f[(0.75, 0.25), (1.25, 0.75), (2.25, 0.75), (1.75, 0.25)]]
)
```

**Comment by jkrumbiegel** [comment]
Yes exactly, it's just that I didn't notice before that the winding order was incorrect because earcut doesn't care, on the mesh conversion route.

We could check winding order and error or flip the polygons in case it's wrong, not sure what the better option is.

**Comment by jaakkor2** [comment]
Checking the winding order is not trivial
```
p = Polygon(
    Point2f[(0, 0), (5, 0), (5, 5), (0, 5)],
    [Point2f[(1, 1), (1, 4), (4, 4), (4, 1)],
     Point2f[(2, 2), (3, 2), (3, 3), (2, 3)]]
)
```
should produce (as it does on CairoMakie + Makie 0.18.3)
![image](https://user-images.githubusercontent.com/7318249/204739463-ba22ccc4-81ef-4f1a-825b-bfec3e7e7e7f.png)

**Comment by fatteneder** [comment]
> Checking the winding order is not trivial
> p = Polygon(
    Point2f[(0, 0), (5, 0), (5, 5), (0, 5)],
    [Point2f[(1, 1), (1, 4), (4, 4), (4, 1)],
     Point2f[(2, 2), (3, 2), (3, 3), (2, 3)]]
)

To me this doesn't look like a correct usage of `poly!`.
Also if you try to add another rectangular cutout inside the small rectangle you would have to type (not tested):
```julia
p = Polygon(
    Point2f[(0, 0), (5, 0), (5, 5), (0, 5)],
    [Point2f[(1, 1), (1, 4), (4, 4), (4, 1)],
     Point2f[(2, 2), (3, 2), (3, 3), (2, 3)],
     Point2f[(2.2,2.2), (2.2,2.8), (2.8,2.8), (2.8,2.2)]]
)
poly!(ax, p)
```

Whereas there is a much simpler way of doing things:
```julia
p1 = Polygon(
    Point2f[(0, 0), (5, 0), (5, 5), (0, 5)],
    [Point2f[(1, 1), (1, 4), (4, 4), (4, 1)]]
)
poly!(ax, p1)
p2 = Polygon(
    Point2f[(2, 2), (3, 2), (3, 3), (2, 3)],
    [Point2f[(2.2,2.2), (2.2,2.8), (2.8,2.8), (2.8,2.2)]]
)
poly!(ax, p2)
```

**Comment by fatteneder** [comment]
> Yes exactly, it's just that I didn't notice before that the winding order was incorrect because earcut doesn't care, on the mesh conversion route.

> We could check winding order and error or flip the polygons in case it's wrong, not sure what the better option is.

Is there anything mentioned about winding order in the docs?
But even if there were, the API still seems not intuitive to me.
E.g. there are only two fields of `Polygon` which are called `exterior` and `interiors` (singular and plural). As it is coded in `CairoMakie`, I think it is always assumed that cutouts are the ones in `interiors`. After all, you cannot even nest a `Polygon` inside the `interiors`, or can you?

To me the signature `poly(exterior, interiors)` suggests that this should draw a polygon that is build from the shape given by `exterior` and every shape given in `interiors` is then subtracted from that.

**Comment by fatteneder** [comment]
> To me the signature poly(exterior, interiors) suggests that this should draw a polygon that is build from the shape given by exterior and every shape given in interiors is then subtracted from that.

If we were to define its usage like that, then there would be no confusion about winding order.

**Comment by jkrumbiegel** [comment]
The reason the nested polygon works in CairoMakie, is that the outlines are just drawn as a multi-segment path, and then filling is determined via the even-odd rule. But I would also argue that this is not canonical polygon behavior. In order to match CairoMakie and GLMakie we should be careful to define what data we expect, so far it has been to loosely defined. We can probably use whatever definitions the leading GIS packages use, as they probably have the most experience with polygons.

**Comment by jaakkor2** [comment]
Sorry for my confusion, I agree my example is malformed
```
p = Polygon(
    Point2f[(0, 0), (5, 0), (5, 5), (0, 5)],
    [Point2f[(1, 1), (1, 4), (4, 4), (4, 1)],
     Point2f[(2, 2), (3, 2), (3, 3), (2, 3)]]
)
```

**Comment by jaakkor2** [comment]
> The reason the nested polygon works in CairoMakie, is that the outlines are just drawn as a multi-segment path, and then filling is determined via the even-odd rule.

The fill rule seems to be `nonzero`
```
using CairoMakie
using Makie.GeometryBasics

f = Figure()
Axis(f[1, 1])

# polygon with hole
p = Polygon(
    Point2f[(0, 0), (2, 0), (3, 1), (1, 1)],
    [Point2f[(0.75, 0.25), (1.75, 0.25), (2.25, 0.75), (1.25, 0.75)]]
)

poly!(p, color = :blue)

fn = "cm.svg"
save(fn, f)
collect(eachmatch(r"fill-rule:(.*?);", read(fn, String)))
```
gives
```
2-element Vector{RegexMatch}:
 RegexMatch("fill-rule:nonzero;", 1="nonzero")
 RegexMatch("fill-rule:nonzero;", 1="nonzero")
```
With `evenodd` fill rule, holes seem to work regardless of winding.

**Comment by ffreyer** [comment]
The docs have been fixed and I think the Polygon discussion is something for GeometryBasics instead... Can this be closed?

**Comment by asinghvi17** [comment]
I would argue not, this is still something we have to figure out in CairoMakie as to whether it should respect this kind of "malformed polygon" or somehow try and correct it before plotting...
